staticd/static_zebra.c

```diff
@@ -530,10 +533,618 @@ extern void static_zebra_route_add(struct static_path *pn, bool install)
+	struct listnode *node;
+	struct static_srv6_sid *sid;
+
+	if (!locator)
+		return -1;
+
+	zlog_info("%s: Received SRv6 locator %s %pFX, loc-block-len=%u, loc-node-len=%u func-len=%u, arg-len=%u",
+		  __func__, locator->name, &locator->prefix, locator->block_bits_length,
+		  locator->node_bits_length, locator->function_bits_length,
+		  locator->argument_bits_length);
+
+	/* If we are already aware about the locator, nothing to do */
+	loc = static_srv6_locator_lookup(locator->name);
+	if (loc)
+		return 0;
+
+	loc = static_srv6_locator_alloc(locator->name);
+
+	zlog_info("SRv6 locator (locator %s, prefix %pFX) set", locator->name, &locator->prefix);
+
+	/* Store the locator prefix */
+	loc->prefix = locator->prefix;
+	loc->block_bits_length = locator->block_bits_length;
+	loc->node_bits_length = locator->node_bits_length;
+	loc->function_bits_length = locator->function_bits_length;
+	loc->argument_bits_length = 0;
+	loc->flags = locator->flags;
+
+	listnode_add(srv6_locators, loc);
+
+	for (ALL_LIST_ELEMENTS_RO(srv6_sids, node, sid)) {
+		if (strncmp(sid->locator_name, loc->name, sizeof(loc->name)) == 0)
+			sid->locator = loc;
+	}
+
+	/* Request SIDs from the locator */
+	request_srv6_sids(loc);
+
+	return 0;
+}
+
+/**
+ * Callback to process an SRv6 locator received from SRv6 Manager (zebra).
+ *
+ * @result 0 on success, -1 otherwise
+ */
+static int static_zebra_process_srv6_locator_add(ZAPI_CALLBACK_ARGS)
+{
+	struct srv6_locator loc = {};
+
+	if (!srv6_locators)
+		return -1;
+
+	/* Decode the SRv6 locator */
+	if (zapi_srv6_locator_decode(zclient->ibuf, &loc) < 0)
+		return -1;
+
+	return static_zebra_process_srv6_locator_internal(&loc);
+}
+
+/**
+ * Callback to process a notification from SRv6 Manager (zebra) of an SRv6
+ * locator deleted.
+ *
+ * @result 0 on success, -1 otherwise
+ */
+static int static_zebra_process_srv6_locator_delete(ZAPI_CALLBACK_ARGS)
+{
+	struct srv6_locator loc = {};
+	struct listnode *node, *nnode;
+	struct listnode *node2, *nnode2;
+	struct static_srv6_sid *sid;
+	struct static_srv6_locator *locator;
+
+	if (!srv6_locators)
+		return -1;
+
+	/* Decode the received zebra message */
+	if (zapi_srv6_locator_decode(zclient->ibuf, &loc) < 0)
+		return -1;
+
+	zlog_info("SRv6 locator deleted in zebra: name %s, "
+		  "prefix %pFX, block_len %u, node_len %u, func_len %u, arg_len %u",
+		  loc.name, &loc.prefix, loc.block_bits_length, loc.node_bits_length,
+		  loc.function_bits_length, loc.argument_bits_length);
+
+	locator = static_srv6_locator_lookup(loc.name);
+	if (!locator)
+		return 0;
+
+	zlog_info("Deleting srv6 sids from locator %s", locator->name);
+
+	/* Delete SRv6 SIDs */
+	for (ALL_LIST_ELEMENTS(srv6_sids, node2, nnode2, sid)) {
+		if (sid->locator != locator)
+			continue;
+
+		zlog_info("Deleting SRv6 SID (locator %s, sid %pFX)", locator->name, &sid->addr);
+
+		/* Uninstall the SRv6 SID from the forwarding plane
+			* through Zebra */
+		static_zebra_srv6_sid_uninstall(sid);
```

Comment:
before calling uninstall(), a check should be done to check that sid has really been provisioned.